### PR TITLE
Fetch block fix

### PR DESCRIPTION
**Pull request by antonlerner**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,7 @@
 module github.com/spacemeshos/go-spacemesh
 
 require (
+	cloud.google.com/go v0.34.0
```
Comment (y0sher) [inline go.mod]: what is this ?

File: mesh/meshdb.go [hunk 2]
```diff
@@ -168,52 +200,35 @@ func (m *meshDB) getLayerBlocks(ids map[BlockID]bool) ([]*Block, error) {
...
 	defer m.lhMutex.Unlock()
-	ll, found := m.layerHandlers[index]
+	ll, found := m.layerMutex[index]
 	if !found {
-		ll = &layerHandler{pendingCount: counter, ch: make(chan *Block)}
-		m.layerHandlers[index] = ll
-		go m.handleLayerBlocks(ll)
-	} else {
-		atomic.AddInt32(&ll.pendingCount, counter)
+		ll = &layerMutex{sync.Mutex{}, sync.Mutex{}, 0}
```
Comment (y0sher) [inline mesh/meshdb.go]: actually, even if you used `&layerMutex{}` you'd get the same values

File: mesh/meshdb.go [hunk 3]
```diff
@@ -168,52 +200,35 @@ func (m *meshDB) getLayerBlocks(ids map[BlockID]bool) ([]*Block, error) {
...
-	if atomic.LoadInt32(&ll.pendingCount) == 0 {
-		delete(m.layerHandlers, ll.layer)
+	defer m.lhMutex.Unlock()
+
+	ll, found := m.layerMutex[index]
+	if !found {
+		panic("trying to double close layer mutex")
 	}
-	m.lhMutex.Unlock()
+
```
Comment (y0sher) [inline mesh/meshdb.go:197]: don't you want to unlock lhMutex here ?

File: mesh/meshdb.go [hunk 4]
```diff
@@ -132,6 +160,10 @@ func (m *meshDB) addLayer(layer *Layer) error {
 }
 
 func (m *meshDB) updateLayerIds(block *Block) error {
```
Comment (y0sher) [inline mesh/meshdb.go]: this func adds a block id layer, it should be named better.

File: mesh/meshdb.go [hunk 5]
```diff
@@ -168,52 +200,35 @@ func (m *meshDB) getLayerBlocks(ids map[BlockID]bool) ([]*Block, error) {
...
-func (m *meshDB) tryDeleteHandler(ll *layerHandler) {
+func (m *meshDB) endLayerWorker(index LayerID) {
 	m.lhMutex.Lock()
-	if atomic.LoadInt32(&ll.pendingCount) == 0 {
-		delete(m.layerHandlers, ll.layer)
+	defer m.lhMutex.Unlock()
+
+	ll, found := m.layerMutex[index]
+	if !found {
+		panic("trying to double close layer mutex")
```
Comment (y0sher) [inline mesh/meshdb.go:195]: sure we want to panic here ?

File: mesh/meshdb.go [hunk 6]
```diff
@@ -74,9 +81,12 @@ func (m *meshDB) addBlock(block *Block) error {
...
 		return errors.New("block " + string(block.ID()) + " already exists in database")
 	}
-
-	layerHandler := m.getLayerHandler(block.LayerIndex, 1)
-	layerHandler.ch <- block
+	err = m.writeBlock(block)
+	if err != nil {
+		return err
+	}
+	/*layerHandler := m.getLayerHandler(block.LayerIndex, 1)
```
Comment (y0sher) [inline mesh/meshdb.go]: remove comments we have history in git :)

File: mesh/meshdb.go [hunk 7]
```diff
@@ -7,7 +7,6 @@ import (
 	"github.com/spacemeshos/go-spacemesh/database"
 	"github.com/spacemeshos/go-spacemesh/log"
 	"sync"
-	"sync/atomic"
 )
 
 type layerHandler struct {
```
Comment (almogdepaz) [inline mesh/meshdb.go]: delete

File: mesh/meshdb.go [hunk 8]
```diff
@@ -16,12 +15,19 @@ type layerHandler struct {
...
+}
+
 type meshDB struct {
 	layers             database.DB
 	blocks             database.DB
 	contextualValidity database.DB //map blockId to contextualValidation state of block
 	orphanBlocks       map[LayerID]map[BlockID]struct{}
 	orphanBlockCount   int32
+	layerMutex         map[LayerID]*layerMutex
 	layerHandlers      map[LayerID]*layerHandler
```
Comment (almogdepaz) [inline mesh/meshdb.go]: delete

File: mesh/meshdb.go (hunk 6 above)
Comment (almogdepaz) [inline mesh/meshdb.go]: ...

File: mesh/meshdb.go [hunk 9]
```diff
@@ -168,52 +200,35 @@ func (m *meshDB) getLayerBlocks(ids map[BlockID]bool) ([]*Block, error) {
...
-		ll = &layerHandler{pendingCount: counter, ch: make(chan *Block)}
-		m.layerHandlers[index] = ll
-		go m.handleLayerBlocks(ll)
-	} else {
-		atomic.AddInt32(&ll.pendingCount, counter)
+		ll = &layerMutex{sync.Mutex{}, sync.Mutex{}, 0}
+		m.layerMutex[index] = ll
 	}
+	ll.pending.Lock()
+	ll.workers++
```
Comment (almogdepaz) [inline mesh/meshdb.go]: change mutex to pendMu or something so you can change workers to pending

**Review by almogdepaz — APPROVED** [review]

**Review by y0sher — APPROVED** [review]

File: mesh/meshdb.go (hunk 5 above)
Comment (antonlerner) [reply mesh/meshdb.go:195]: yep, shouldn't happen here

File: mesh/meshdb.go (hunk 3 above)
Comment (antonlerner) [reply mesh/meshdb.go:197]: fixed

**Pull request merged**